### ISSUE TITLE
Syntax sugar for "in place" declaration of variables passed to ByRef parameters.

### ISSUE THREAD

**Issue by ericmutta** [dotnet/vblang#159]
@AnthonyDGreen  this is a follow up to #152 where I asked if we'd get full support for `out` vars and said it would make TryXXX methods easier. What I was getting at is some syntax sugar, so for example this:

```vb
If Integer.TryParse("123", Dim num) Then
 Console.WriteLine($"Got number {num}")
End If
```

Would be transformed to this:

```vb
Dim num as Integer
If Integer.TryParse("123", num) Then
 Console.WriteLine($"Got number {num}")
End If
```

It was that aspect of `out` vars that jumped out at me when I first heard about them. 

Issues like definite assignment and other related complexities never came to mind since VB has never checked those things as an inherent part of the language (in fact I often leverage such quirks for example, to implicitly return `false` in a function where the `true` path may not always execute...the IDE usually says "warning, code may not return a value on all paths" to which I always say, "I know but VB works a little different" 😄).

**Comment by reduckted** [comment]
This is basically #60, right?

**Comment by ericmutta** [comment]
> @reduckted This is basically #60, right?

Yes, it's exactly the same thing. Looks like this has been in the air for a few months in various forms (#60 and #132). I propose we put our heads together and start hashing out a more formal spec. Here is a starting point:

1. An inline variable declaration can appear in any location where an expression is allowed.

2. If the declaration is of the form `Dim v` or `Dim v As T` then it must appear in a method call and must be in a position corresponding to a ByRef parameter `P` of the called method.

    **A)** For `Dim v` the type of the inline variable `v` is inferred to be the type of the corresponding ByRef parameter `P` in the called method.

    **B)** For `Dim v As T` the type of the inline variable `v` is `T`. A compile-time error occurs if there's no implicit conversion from the type of the ByRef parameter `P` to the type `T`.

3. If the declaration is of the form `Dim v = E` or `Dim v As T = E` then it can appear anywhere the expression `E` is allowed to appear by itself.

    **A)** For `Dim v = E` the type of the inline variable `v` is inferred to be the type of the expression `E`.

    **B)** For `Dim v As T = E` the type of the inline variable `v` is `T`. A compile-time error occurs if there's no implicit conversion from the type of the expression `E` to the type `T`.

Now for some code examples.

Rule 2A covers the `Integer.TryParse()` and related scenarios we've already been talking about, so no code example is needed here.

Rule 2B exists for completeness and allows you to use an explicit type and addresses the type conversion issues:

```vb
If Integer.TryParse("123", Dim num As Long) Then Console.WriteLine(num)

'the above is implemented as follows:

Dim temp_int as Integer
Dim temp_bool As Boolean = Integer.TryParse("123", temp_int)
Dim num As Long = temp_int
If temp_bool Then Console.WriteLine(num)
```

**Rule number 3 however is new and allows the following scenarios**

Usage in an If-statement:

```vb
If (Dim v = GetNumber()) > 10 Then PrintNumber(v)

'the above is equivalent to:

Dim v = GetNumber()
If v > 10 Then PrintNumber(v)
```

Usage inside interpolated strings:

```vb
Console.WriteLine("Enter your name: ")
Console.WriteLine($"Hello {Dim name = Console.ReadLine()}. Please wait...")
DoSomething(name)

'the above is equivalent to:

Console.WriteLine("Enter your name: ")
Dim name = Console.ReadLine()
Console.WriteLine($"Hello {name}. Please wait...")
DoSomething(name)
```

Usage in `While` loops (note the `For` and `For Each` loops already allow inline declaration).

```vb
Do While (Dim line = MyStream.ReadLine()) <> ""
 ProcessLine(line)
Loop

'the above is equivalent to:

Do
  Dim line = MyStream.ReadLine()
  If line <> "" Then
    ProcessLine(line)
  Else
    Exit Do
  End If
Loop
```

Usage in a `Select Case`

```vb
Select Case True
 Case Integer.TryParse(MyInput, Dim TheNumber)
   ProcessNumber(TheNumber)
 Case Date.TryParse(MyInput, Dim TheDate)
   ProcessDate(TheDate)
End Select

'the above is equivalent to:

Dim TheNumber As Integer
Dim TheDate As Date
Select Case True
 Case Integer.TryParse(MyInput, TheNumber)
   ProcessNumber(TheNumber)
 Case Date.TryParse(MyInput, TheDate)
   ProcessDate(TheDate)
End Select
```

Usage in a `Select Case` with lazy evaluation:

```vb
Select Case True
  Case (Dim n = GetNum()) > 10
     PrintNumber(n)
  Case (Dim d = GetDate()) < Date.Now
     PrintDate(d)
End Select

'the above is equivalent to:

Dim temp_num = GetNum()
If temp_num > 10 Then
  Dim n = temp_num
  PrintNumber(n)
Else
  Dim temp_date = GetDate()
  If temp_date < Date.Now Then
    Dim d = temp_date
    PrintDate(d)
  End If
End If
```

Usage in variable declarations with initialisation (this looks a bit weird but rule 3 allows it):

```vb
Dim one = (Dim two = (Dim three = 1000))

'is equivalent to:

Dim three = 1000
Dim two = three
Dim one = two
```

I think that covers each major area (assignment statements, invocation statements, conditional statements and loop statements).

**This proposal is actually an expansion of an existing language feature to new areas.** It is has been possible to do the following for years now in VB:

```vb
For n As Integer = 1 To 10
Next
```

Note also that this feature effectively allows assignment statements _inside_ expressions (C# already allows this because its assignment and equality operators are distinct so there's never ambiguity).  This feels like a new thing for VB but it is not: the `Let` keyword in LINQ has always been able to do inline declarations with assignment inside expressions:

```vb
    Dim OddNumbers = From n In Enumerable.Range(1, 10)
                     Let IsOdd = n Mod 2 <> 0
                     Where IsOdd
                     Select n 
```

...so we do have some history to work with. In any case, that's a starting point, I am sure others can pitch in with concerns/enhancements! 

CC: @AnthonyDGreen

**Comment by craigajohnson** [comment]
@ericmutta this is intriguing.

Could this same concept possibly be expanded to include a long-standing request to be able to name the subject of a With block and then reference that name within the block?

For instance:

```
With (Dim x = Context.CreateGrover("Quip"))
   .Add("Monkey")
   .Add("Turnip")
   .Process(5, 2, "Arrows")
   NowPassThisBeast(x)
End With
```

The existing request was to allow `With x = Context.CreateGrover("Quip")` and then reference x. However, I don't think the additional Dim gets in the way, and, if anything, makes the intent even more clear.

**Comment by ericmutta** [comment]
> @craigajohnson Could this same concept possibly be expanded to include a long-standing request to be able to name the subject of a With block and then reference that name within the block?

Absolutely! In fact I had forgotten about `With...End With` but rule number 3 covers the scenario without any extensions because it says this:

```vb
With (Dim x = Context.CreateGrover("Quip"))
   .Add("Monkey")
   .Add("Turnip")
   .Process(5, 2, "Arrows")
   NowPassThisBeast(x)
End With
```

Simply becomes this:

```vb
Dim x = Context.CreateGrover("Quip")

With x
   .Add("Monkey")
   .Add("Turnip")
   .Process(5, 2, "Arrows")
   NowPassThisBeast(x)
End With
```

Also, those brackets could be removed in cases where you are not checking the value of the inline variable:

```vb
With Dim x = GetList()  '<---- no brackets needed here.
End With

'no brackets needed below:

If Dim MyBoolean = GetBoolean() Then
End If
```

**Comment by ericmutta** [comment]
> @KathleenDollard So if there is something you want reviewed early, please make a comment - ideally indicating why you think it should be reviewed (quoted from #251).

My second vote for early review goes to this issue. As for why:

1) It has appeared in various forms (see earlier comments). 

2) It hasn't received any major objections from anyone.

3) I hashed out a rough spec above with syntax and semantics of generated code.

4) It has wide applicability (see various scenarios in earlier comments).

5) It is "minor change" in the sense that no new keywords or operators are being added, nothing needs to be changed in the CLR, etc.

**Comment by jrmoreno1** [comment]
Either this or 172 would have been great for some code I am looking at today..
